### DISCUSSION TITLE
Engine Team Organization

### DISCUSSION THREAD

**Discussion by CharliePoole** [nunit/nunit-console#1523] [Team Discussions]
Welcome to this new discussion category, intended primarily for discussions within the engine team but also open to the public.

With the recent additions of @veleek, @manfred-brands and @stevenaw, the engine team now has [8 members](https://github.com/orgs/nunit/teams/engine-team), some quite active, others less so.

Of course, that's to be expected. Folks join a team for various reasons and both their interests and their ability to participate tend to vary over time. Some typical ways of participating include...

- Writing code to fix bugs, make improvements or add features.
- Reviewing code written by others.
- Participating in discussions about the future direction of the engine and console.
- Answering questions for users.
- Just listening in so as to keep informed.

It would be helpful to me if each team member would let me know what level of participation you are looking forward to over the next six months or so. You can post here or send me an email, as you prefer.

In particular, unless you've already recently told me, I'd like to know who is willing to do code reviews. I plan to set up automatic assignment of reviews in GitHub. I haven't used that before and I'm interested to see how well it works. 

Also, if you're a team member whose email I don't have and it's not visible in GitHub, please tell me how to contact you directly.

**Comment by manfred-brands** [comment]
@CharliePoole I'm happy to review and to have a sound-board for your ideas.  

I'm hesitant to make big changes to code that is not nullable enabled or in code that has not enabled the MS recommend analyzers. Ever since we introduced those in our company code has improved. Passing null to a method not expecting it is a thing of the past. Converting existing code to nullable is a task on its own especially if it uses fields in a method which it assumes are set beforehand. Declaring these fields a nullable would result in a lot of unnecessary tests in code unless we can use `MemberNotNull` or update the code not to modify state but to expect parameters and return update results.
If the test coverage is good, then enabling nullability before starting on changes for V4 is something I would recommend.
If you prefer to leave it, that is your decision as team-lead.

**Comment by CharliePoole** [comment]
@manfred-brands That's great. And hopefully you can sound off with some of your own ideas as well.

I agree with you that the code needs to be null-enabled. I haven't wanted to do it in the V3 code base but I agree that it ought to be done with V4. We'll need to do some planning about this. For example, I'm not sure whether to do do it immediately or after porting relevant changes from 3.18 and 3.19 to the version4 branch.

**Comment by manfred-brands** [comment]
How much does the V4 branch differ or could we start from the latest V3?

**Comment by CharliePoole** [comment]
They differ quite a bit. Most We put quite a bit of work into it a few years ago and many of the biggest breaking changes are already in place. I believe most improvements through 3.16 have been ported to it but not the latest releases. Maybe making the change to null-enabled first would be the better strategy, since ported fixes would then give error messages we could fix.  The advantage of that is it could be started sooner if somebody wanted to give it a shot.

**Comment by mikkelbu** [comment]
I think I'll only be able to participate on a smaller scale - ideally I would want to help a lot, but this would just be an empty hope/promise as I don't have much spare time at the moment. So my guess I would mostly be doing smaller tasks - but not confined to a particular category of tasks (even just keeping up with the flow takes some portion of my time).

If we wanted to get an idea about the difference between v3 and v4 is it then comparing the `main` and `version4` branches ?

**Comment by CharliePoole** [comment]
@mikkelbu There will be small tasks, for sure. I'll be starting a thread soon with more information on V4 but looking at the completed issues and prs in the 4.0 milestone is probably the best way to start.

**Comment by stevenaw** [comment]
Thanks for inviting me @CharliePoole 

Spare time is unpredictable at the moment. As I'm newer on the console side than other areas of NUnit I would be looking to at first to take the opportunity to learn more through code reviews and answering questions in issues. I'm already following along in the issue discussions :) Over time I'd hope to then get more involved in coding as time allows and as I build up a better working knowledge of historical issues and approach to how things have been done within the console thus far.
